specification/transport/udp/udp.tex

```diff
@@ -0,0 +1,337 @@
+}{|l l l l X|}
+    Field & Offset & Width & Value & Description \\
+
+    RFC~1112 multicast prefix &
+    28 & 4 & $1110_2$ &
+    \\
+
+    RFC~2365 scope &
+    24 & 4 & $1111_2$ &
+    Selects the administratively scoped range 239.0.0.0/8 per RFC~2365
+    to avoid collisions with well-known multicast groups. \\
+
+    RFC~2365 reserved range &
+    23 & 1 & $0_2$ &
+    Selects the ad-hoc defined range 239.0.0.0/9 per RFC~2365. \\
+
+    Cyphal/UDP address version &
+    22 & 1 & $0_2$ &
+    Deconflicts this layout with future revisions. \\
+
+    Reserved &
+    17 & 5 & $00000_2$ &
+    May be used for domain-ID segregation in future versions. \\
+
+    Z: service, not message &
+    16 & 1 & any &
+    Set for service transfers, cleared for message transfers. \\
+
+    X if Z: destination node-ID &
+    0 & 16 & $[0, 65534]$ &
+    The destination node-ID of the current service transfer. \\
+
+    X if not Z: subject-ID &
+    0 & 16 & $[0, 8191]$ &
+    The subject-ID of the current message transfer. \\
+\end{CyphalSimpleTable}
+
+A subscriber to certain Cyphal subjects will join the IP multicast groups corresponding to said subjects.
+Likewise, a node that provides at least one RPC-service will join the IP multicast group corresponding to
+its own node-ID\footnote{Observe that multicast groups are not differentiated by service-ID.}.
+
+The IP address of a node bears no relevance for the protocol ---
+multiple nodes may share the same IP address; likewise, a node may have more than one IP address.
+Nodes on a Cyphal/UDP network are identified exclusively by their node-ID value\footnote{%
+    A node that is registered on an IP network (e.g., via DHCP)
+    still needs to obtain a node-ID value to participate in a Cyphal/UDP network.
+    This may be done either through manual assignment or by using the plug-and-play node-ID allocation service
+    (section~\ref{sec:application_functions_pnp}).
+}.
+The set of valid node-ID values for Cyphal/UDP is $[0, 65534]$.
+Value 65535 is reserved to represent both the broadcast and anonymous node-ID, depending on context.
+
+Sources of Cyphal/UDP traffic should set the packet TTL to 16 or higher\footnote{%
+    RFC~1112 prescribes a default TTL of 1,
+    but this is not sufficient for most applications Cyphal/UDP is intended for.
+}.
+
+The DSCP\footnote{RFC~2474} field of outgoing IP packets
+should be populated based on the Cyphal transfer priority level (section~\ref{sec:transport_transfer_priority})
+such that the maximum Cyphal priority level corresponds to class selector CS7
+and the minimum Cyphal priority level corresponds to class selector CS0,
+with the intermediate values mapped following the same principle.
+
+\begin{remark}
+    \begin{CyphalSimpleTable}{
+        Recommended DSCP class selector values\label{table:transport_udp_priority}
+    }{|l l l l|}
+        Cyphal priority & DSCP class selector & IP precedence value & IP precedence name      \\
```

Comment:
The reason I went with my original suggestion is the expectation that intra-vehicular routing policies will likely be configured from scratch and may not inherit the policies suggested in certain RFCs. However, this is not to say that consistency with COTS has no value.

The general principle behind the original IP precedence field is that higher values correspond to higher priorities (RFC791), which is upheld for reasons of backward compatibility in the DSCP field (RFC2474). Then, RFC3662 introduces the LE policy with the suggestion to mark it as DSCP CS1, but unless I am missing something, it makes no statement that CS1 should be treated as a lower-priority class compared to CS0, and indeed that would violate RFC791 and RFC1347 (the relevant part is *"If a host makes appropriate use of the TOS facility, its network service should be at least as good as (and hopefully better than) it would have been if the host had not used the facility"*). Therefore the DSCP values you suggested for the Low and Optional priority levels appear incorrect. Would you agree?

If we agree that the general principle of higher precedence corresponding to higher priority is likely to be upheld in all networks we care about, we could just as well keep the simple original mapping I initially proposed. Otherwise, we need to revise the lower levels.

Reply:
I think we should bias towards enabling no-config behavioiur that is optimal for Cyphal/UDP and develop Cyphal/TSN as the protocol one should use when tuning a vehicle network to be optimal and deterministic. For UDP we should look to switches like [this](https://www.amazon.com/Ethernet-Splitter-Optimization-Unmanaged-TL-SG105/dp/B00A128S24) where IP telephony is pre-configured. So, actually, let me modify my proposal to exclude the AF41 and just support three types: CS1/0/lowest == optional (it does seem that this is actually lower than 1), DF/0/default == low, and EF/46/High == nominal.

Finally, note that value 7 is actually reserved for future use and may not be routed on some networks so a flat mapping is both wrong (i.e. 1 and 0 may be inverted in priority) and illegal (in that 7 is reserved).

Reply:
I have updated this part as discussed on Matrix: https://github.com/OpenCyphal/specification/pull/127/commits/1c33eaf32d03a769cec02bdc066f4e2b40a588dc. The specific level mapping may require bikeshedding but we need to keep in mind that per RFC8837, "excess" EF traffic is dropped. We may want to consider replacing EF with AF41.

Reply:
Indeed, it is hard to come up with a sensible recommendation given the uncertainty. I will update this to require zero by default with configurability.